presto-main/src/main/java/com/facebook/presto/util/FastutilSetHelper.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.util;
+
+import com.google.common.collect.ImmutableMap;
+import it.unimi.dsi.fastutil.booleans.BooleanOpenHashSet;
+import it.unimi.dsi.fastutil.bytes.ByteOpenHashSet;
+import it.unimi.dsi.fastutil.chars.CharOpenHashSet;
+import it.unimi.dsi.fastutil.doubles.DoubleOpenHashSet;
+import it.unimi.dsi.fastutil.floats.FloatOpenHashSet;
+import it.unimi.dsi.fastutil.ints.IntOpenHashSet;
+import it.unimi.dsi.fastutil.longs.LongOpenHashSet;
+import it.unimi.dsi.fastutil.objects.ObjectOpenHashSet;
+import it.unimi.dsi.fastutil.shorts.ShortOpenHashSet;
+
+import java.lang.reflect.InvocationTargetException;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Set;
+
+import static com.google.common.base.Preconditions.checkState;
+
+public final class FastutilSetHelper
+{
+    private static final Map<Class<?>, Class<? extends Set<?>>> TYPE_TO_TYPED_HASH_SET = ImmutableMap.<Class<?>, Class<? extends Set<?>>>builder()
+            .put(boolean.class, BooleanOpenHashSet.class)
+            .put(byte.class, ByteOpenHashSet.class)
+            .put(char.class, CharOpenHashSet.class)
+            .put(double.class, DoubleOpenHashSet.class)
+            .put(float.class, FloatOpenHashSet.class)
+            .put(int.class, IntOpenHashSet.class)
+            .put(long.class, LongOpenHashSet.class)
+            .put(short.class, ShortOpenHashSet.class)
+            .build();
+
+    public static <T> Set<T> toFastutilHashSet(Set<T> set, Class<T> javaElementType)
+    {
+        Class<? extends Set<T>> fastutilHashSetClass = fastutilHashSetClassFor(javaElementType);
+        try {
+            return fastutilHashSetClass.getConstructor(Collection.class).newInstance(set);
```

Comment:
I have updated & rebased the code.

I modified the function to use branched if. I agree that it looks better.

> At least, it's checking that what you pass to the constructor is a collection at compile time.

Does it make a difference in this case? We know that argument is `Set` anyway and the cost is negligible.

> Btw, I think the signature for that method should be:

I changed the signature since indeed `T` is different for `javaElementType` and `set` parameters (one is a primitive type class, the other is boxed type class)